pkg/controller/worker/state_reconciler.go

```diff
@@ -0,0 +1,140 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package worker
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gardener/gardener-extensions/pkg/controller"
+	extensionscontroller "github.com/gardener/gardener-extensions/pkg/controller"
+	"github.com/gardener/gardener-extensions/pkg/util"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	gardencorev1beta1helper "github.com/gardener/gardener/pkg/apis/core/v1beta1/helper"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+
+	"github.com/go-logr/logr"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/util/retry"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+)
+
+type stateReconciler struct {
+	logger   logr.Logger
+	actuator StateActuator
+
+	ctx    context.Context
+	client client.Client
+}
+
+// NewStateReconciler creates a new reconcile.Reconciler that reconciles
+// Worker's State resources of Gardener's `extensions.gardener.cloud` API group.
+func NewStateReconciler(mgr manager.Manager, actuator StateActuator) reconcile.Reconciler {
+	return &stateReconciler{
+		logger:   log.Log.WithName(StateUpdatingControllerName),
+		actuator: actuator,
+	}
+}
+
+func (r *stateReconciler) InjectFunc(f inject.Func) error {
+	return f(r.actuator)
+}
+
+func (r *stateReconciler) InjectClient(client client.Client) error {
+	r.client = client
+	return nil
+}
+
+func (r *stateReconciler) InjectStopChannel(stopCh <-chan struct{}) error {
+	r.ctx = util.ContextFromStopChannel(stopCh)
+	return nil
+}
+
+func (r *stateReconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	worker := &extensionsv1alpha1.Worker{}
+	if err := r.client.Get(r.ctx, request.NamespacedName, worker); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Deletion flow
+	if worker.DeletionTimestamp != nil {
+		//Nothing to do
+		return reconcile.Result{}, nil
+	}
+
+	// Reconcile flow
+	if err := controller.EnsureFinalizer(r.ctx, r.client, FinalizerName, worker); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	operationType := gardencorev1beta1helper.ComputeOperationType(worker.ObjectMeta, worker.Status.LastOperation)
+	if operationType == gardencorev1beta1.LastOperationTypeCreate {
+		//Something which is not created does not need state update for the moment so we leave it for later
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	if err := r.updateStatusProcessing(r.ctx, worker, operationType, "Updateting the worker state"); err != nil {
```

Comment:
```suggestion
	if err := r.updateStatusProcessing(r.ctx, worker, operationType, "Updating the worker state"); err != nil {
```

Reply:
But wait a sec, why do you need to update the `lastOperation` here in the first place? The `lastOperation` should only be touched by the "real"/main controller. The state updater should only update the state.

Reply:
I was wondering does this auxiliary controller has to left trace of his work in the Worker state and if so, where?

Reply:
Hm, what about sending `Events`?

Reply:
This is good Idea. I will implement it.